### ISSUE TITLE
[RFC] Deterministic Reward Inference and VLM Reward Model Support

### ISSUE THREAD

**Issue by KaisennHu** [verl-project/verl-omni#111]
## Motivation

verl-omni's reward framework supports two reward model pathways:

- **DisRM** (discriminative): text-only, subject to floating-point nondeterminism, **cannot handle visual input**
- **GenRM** (generative): supports VLM Vision API, but **sampling params hardcoded** (`temperature=0.7, top_p=0.8`) in every `/v1/chat/completions` request, with no seed control for reproducibility

Three gaps:

1. **No floating-point determinism** — flash attention split-K, CUDA matmul, NCCL all-reduce, MoE routing can all produce floating-point variation across RM inference runs
2. **No configurable GenRM sampling** — GenRM's hardcoded per-request params bypass RM rollout config; users cannot control sampling behavior or provide seed for reproducibility
3. **No VLM scoring path** — enabling RM without a custom reward function has no built-in visual scoring entry point, requiring manual configuration

## Design

Add `deterministic` and `seed` fields under `reward_model` in YAML config:

- **`deterministic=true`** — full determinism applied to RM actors via `enable_full_determinism(seed)` + `PYTHONHASHSEED` injected via `runtime_env`; `seed` propagated to GenRM HTTP requests (same seed + same input → same output, controlled pseudorandom)
- **`deterministic=false`** — full randomness, `seed` not propagated
- **Sampling params** (`temperature/do_sample/top_k/top_p`) are entirely user-configured in YAML under `rollout`, the `deterministic` flag does not override them

### 1. Floating-point determinism

RM inference uses `name=vllm` servers (`vLLMReplica` + `vLLMHttpServer`). When `deterministic=true`, monkey-patch `vLLMReplica.launch_servers()` with `_DeterministicServerProxy` which does two things:

**Env vars injection** — `PYTHONHASHSEED` must be set before Python process startup, so it is injected into RM actor `runtime_env`. Rollout actors remain unaffected (`is_reward_model` guard).

**Server class swap** — Replace `vLLMHttpServer` with `_DeterministicRMHttpServer(vLLMHttpServer)` subclass, which calls verl's `enable_full_determinism(seed)` in `_post_init`. This covers all in-process determinism: env vars (`CUBLAS_WORKSPACE_CONFIG`, `FLASH_ATTENTION_DETERMINISTIC`, `VLLM_BATCH_INVARIANT`, `NCCL_LAUNCH_MODE`, `NCCL_PROTO`), Python/NumPy/PyTorch seeds, deterministic algorithms, cuDNN config.

| Setting | Set by |
|---|---|
| `PYTHONHASHSEED=str(seed)` | `runtime_env` (before process startup) |
| `CUBLAS_WORKSPACE_CONFIG=:16:8` | `enable_full_determinism` (in-process) |
| `FLASH_ATTENTION_DETERMINISTIC=1` | `enable_full_determinism` (in-process) |
| `VLLM_BATCH_INVARIANT=1` | `enable_full_determinism` (in-process) |
| `NCCL_LAUNCH_MODE=GROUP` | `enable_full_determinism` (in-process) |
| `NCCL_PROTO=Simple` | `enable_full_determinism` (in-process) |
| `random.seed(seed)` | `enable_full_determinism` (in-process) |
| `np.random.seed(seed)` | `enable_full_determinism` (in-process) |
| `torch.manual_seed(seed)` | `enable_full_determinism` (in-process) |
| `torch.cuda.manual_seed_all(seed)` | `enable_full_determinism` (in-process) |
| `torch.use_deterministic_algorithms(True, warn_only=True)` | `enable_full_determinism` (in-process) |
| `cudnn.deterministic=True, benchmark=False, enabled=False` | `enable_full_determinism` (in-process) |

### 2. Configurable GenRM sampling + seed control

`VisualRewardManager` passes RM rollout config sampling params to `compute_score_ocr` via `sampling_params` kwarg, replacing hardcoded `DEFAULT_SAMPLING_PARAMS`. When `deterministic=true`, `seed` (from `reward_model.seed`) is included in `sampling_params` and propagates to vLLM's `/v1/chat/completions` request. When `deterministic=false`, `seed` is excluded — full randomness.

Sampling params (`temperature/do_sample/top_k/top_p`) are user-configured. Examples:
- `temperature=0, top_p=1.0` → greedy, fully deterministic
- `temperature>0, seed=42` → controlled pseudorandom, reproducible

### 3. VLM scoring path

When RM enabled + no `custom_reward_function.path`, auto-set path to `compute_score_ocr`. This provides the GenRM scoring entry point needed for VLM visual scoring.

### Propagation chain

```
reward.yaml (deterministic: true/false, seed: 42, rollout.temperature/do_sample/top_k/top_p)
  → DiffusionRayTrainer._init_online_rollout_stack()
      ├─ deterministic=true: monkey-patch vLLMReplica → _DeterministicServerProxy
      │     ├─ inject PYTHONHASHSEED into runtime_env
      │     └─ swap server_class → _DeterministicRMHttpServer (calls enable_full_determinism in _post_init)
      ├─ RM enabled + no custom_reward_function: auto-set → compute_score_ocr
  → VisualRewardManager
      ├─ sampling_params from rollout config (temperature, do_sample, top_k, top_p)
      ├─ deterministic=true: include seed from reward_model.seed
      ├─ deterministic=false: exclude seed
      → compute_score_ocr(sampling_params=...)
          → HTTP /v1/chat/completions request uses user-configured params
```

---

### 1. YAML config

**File**: `verl_omni/trainer/config/reward/reward.yaml`

Add `deterministic: false` and `seed: 42` under `reward_model`.

### 2. Floating-point determinism

**File**: `verl_omni/trainer/diffusion/ray_diffusion_trainer.py`

Define `_DeterministicRMHttpServer(vLLMHttpServer)` subclass. When `deterministic=true` + RM enabled, monkey-patch `vLLMReplica` with `_reward_deterministic` class attribute and `launch_servers()` override via `_DeterministicServerProxy`. `_DeterministicServerProxy` injects `PYTHONHASHSEED` into RM actor `runtime_env` and swaps `self.server_class` to `_DeterministicRMHttpServer`.

**File**: `verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py`

Remove `_DETERMINISTIC_RM_ENV_VARS` static dict.

### 3. Configurable GenRM sampling + seed control

**File**: `verl_omni/reward_loop/reward_manager/visual.py`

Pass `sampling_params` from RM rollout config. Include `seed` when `deterministic=true`, exclude when `deterministic=false`.

**File**: `verl_omni/utils/reward_score/genrm_ocr.py`

Add `sampling_params: dict = None` parameter. Use `sampling_params or DEFAULT_SAMPLING_PARAMS` in HTTP request.

---

## Backward Compatibility

- `deterministic=true` affects RM actors only; rollout unaffected
- `deterministic=false` preserves existing behavior
- `sampling_params` defaults to `None`; when `None`, `DEFAULT_SAMPLING_PARAMS` used
- `custom_reward_function` auto-set when RM enabled + no path provided

## Verification

1. `deterministic=true`: `PYTHONHASHSEED` injected via runtime_env + `enable_full_determinism(seed)` called in `_post_init`
2. Rollout actors unaffected (`is_reward_model` guard)
3. `deterministic=true` + `seed=42`: reproducible GenRM scores
4. `deterministic=false`: existing behavior

## Files

| File | Change |
|---|---|
| `verl_omni/trainer/config/reward/reward.yaml` | Add `deterministic: false` and `seed: 42` under `reward_model` |
| `verl_omni/trainer/diffusion/ray_diffusion_trainer.py` | `_DeterministicRMHttpServer` + monkey-patch `vLLMReplica` |
| `verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py` | Remove `_DETERMINISTIC_RM_ENV_VARS` |
| `verl_omni/reward_loop/reward_manager/visual.py` | Pass `sampling_params`; include/exclude `seed` |
| `verl_omni/utils/reward_score/genrm_ocr.py` | Add `sampling_params` param |
| `tests/reward_loop/test_visual_reward_manager.py` | `test_deterministic_reward_reproducibility` |

## Related Issues
Closes #97

## CC Lists
@SamitHuang @zhtmike

**Comment by SamitHuang** [comment]
For "Sampling-level: Override rollout sampling params to greedy (temperature=0, do_sample=false) via override_generation_config propagation", it may break the GRPO training since GRPO relies on stochastic exploration during rollout. So we should not modify `RolloutConfig` in `ray_diffusion_trainer.py`

Do you mean making the reward model inference to be greedy?

**Comment by zhtmike** [comment]
> For "Sampling-level: Override rollout sampling params to greedy (temperature=0, do_sample=false) via override_generation_config propagation", it may break the GRPO training since GRPO relies on stochastic exploration during rollout. So we should not modify `RolloutConfig` in `ray_diffusion_trainer.py`
> 
> Do you mean making the reward model inference to be greedy?

It is for reward model. A deterministic return for reward model is fine. We need stochastic for rollout instead of reward.

**Comment by shizuku-desir7e** [comment]
Hello, I’m a zero-year master’s student at Sun Yat-sen University (SYSU), currently learning RL and studying the Verl-Omni codebase. I’m especially interested in the reward inference part. If this direction aligns with the project roadmap, I’d like to take a shot at implementing this PR. My understanding is that the problem mainly has two parts: first, adding a deterministic switch for Reward Model inference that applies only to the reward path, so it does not interfere with the stochastic exploration behavior of rollout/policy sampling; second, adding a built-in VLM reward scoring path to the existing VisualRewardManager for image-based reward evaluation. My initial plan is to introduce a reward_model.deterministic config option so that, when enabled, reward requests use greedy decoding and deterministic-related environment variables are injected only into reward-serving workers, while rollout sampling remains completely unchanged. For the visual reward part, I plan to add a built-in image + scoring prompt request path when no custom_reward_function is provided, and parse a scalar reward score from the VLM’s text response. To reduce risk, I would also prefer to submit this in stages: first the RM-only deterministic plumbing, then configurable GenRM sampling parameters, and finally the built-in VLM reward scoring path. If the maintainers think this direction makes sense, I’d be happy to start with a small-scope PR or a draft PR for review. I’d also really appreciate any feedback if there are flaws in my proposal, and I’m happy to learn and iterate as I work on it.

**Comment by zhtmike** [comment]
cc @MmMaiIIi

**Comment by KaisennHu** [comment]
@SamitHuang @zhtmike Clarified that deterministic configuration only affects `config.reward.reward_model.rollout`. Please take another look! Thanks.
